massa-network-worker/src/node_worker.rs

```diff
@@ -65,111 +62,52 @@ impl NodeWorker {
-                    self.node_id
-                );
-                Ok(())
-            }
-            Err(TrySendError::Closed(_)) => {
-                debug!("failed sending message deconnected {}.", self.node_id);
-                Err(NetworkError::ChannelError(
-                    "failed sending message to node: channel closed".into(),
-                ))
-            }
-            Ok(_) => Ok(()),
-        }
-    }
-
     /// node event loop. Consumes self.
     pub async fn run_loop(mut self) -> Result<ConnectionClosureReason, NetworkError> {
-        let (writer_command_tx, mut writer_command_rx) =
-            mpsc::channel::<Message>(NODE_SEND_CHANNEL_SIZE);
         let mut socket_writer = self.socket_writer_opt.take().ok_or_else(|| {
             NetworkError::GeneralProtocolError(
                 "NodeWorker call run_loop more than once".to_string(),
             )
         })?;
-        let write_timeout = self.cfg.message_timeout;
-        let node_id_copy = self.node_id;
+
         let node_writer_handle = tokio::spawn(async move {
-            loop {
-                match writer_command_rx.recv().await {
-                    Some(to_send) => {
-                        match timeout(write_timeout.to_duration(), socket_writer.send(&to_send))
-                            .await
-                        {
-                            Err(_err) => {
-                                massa_trace!("node_worker.run_loop.loop.writer_command_rx.recv.send.timeout", {
-                                    "node": node_id_copy,
-                                });
-                                return Err(std::io::Error::new(
-                                    std::io::ErrorKind::TimedOut,
-                                    "node data writing timed out",
-                                )
-                                .into());
-                            }
-                            Ok(Err(err)) => {
-                                massa_trace!("node_worker.run_loop.loop.writer_command_rx.recv.send.error", {
-                                    "node": node_id_copy, "err":  format!("{}", err),
-                                });
-                                return Err(err);
-                            }
-                            Ok(Ok(id)) => {
-                                massa_trace!("node_worker.run_loop.loop.writer_command_rx.recv.send.ok", {
-                                    "node": node_id_copy, "msg_id": id,
-                                })
-                            }
-                        }
-                    }
-                    None => {
-                        massa_trace!("node_worker.run_loop.loop.writer_command_rx.recv. None", {});
-                        break;
-                    }
-                };
-            }
-            Ok(())
+            node_writer_handle(
+                &mut socket_writer,
+                &mut self.node_command_rx,
```

Comment:
Is it legal to spawn a task with a reference to self.node_command_rx ?!
To make things clean, I would use an `Option<Receiver>` for self.node_command_rx and take() the receiver just before the spawn. Then it would be moved to the task which would be its sole owner

Reply:
It compiles and seems to work fine but that would be interesting to have a deeper look and understand this...